src/Lean/AddDecl.lean

```diff
@@ -68,20 +68,55 @@ def wasOriginallyTheorem (env : Environment) (declName : Name) : Bool :=
   getOriginalConstKind? env declName |>.map (· matches .thm) |>.getD false
 
 /-- If `warn.sorry` is set to true, then, so long as the message log does not already have any errors,
-declarations with `sorryAx` generate the "declaration uses 'sorry'" warning. -/
+declarations with `sorryAx` generate the "declaration uses `sorry`" warning. -/
 register_builtin_option warn.sorry : Bool := {
   defValue := true
   descr    := "warn about uses of `sorry` in declarations added to the environment"
 }
 
+/-- Returns true if a name component is an internal auxiliary suffix that should
+be stripped for user-facing messages. Uses the same criteria as `Name.isInternalDetail`
+for leaf components: names starting with `_`, or matching patterns like `proof_N`, `eq_N`, etc. -/
+private def isAuxiliarySuffix (s : String) : Bool :=
+  s.startsWith "_" ||
+  matchPrefix s "eq_" ||
+  matchPrefix s "match_" ||
+  matchPrefix s "proof_" ||
+  matchPrefix s "omega_"
+where
+  /-- Check that a string begins with the given prefix, and then is only digits/'_'. -/
+  matchPrefix (s : String) (pre : String) :=
+    s.startsWith pre && (s.drop pre.length).all fun c => c.isDigit || c == '_'
+
+/-- Get displayable names for a declaration, for use in sorry warnings.
+Strips auxiliary suffixes from the leaf component so users see the original declaration name.
+For anonymous declarations like `_example`, displays as `example`. -/
+def Declaration.namesForSorry (decl : Declaration) : List Name :=
+  let stripAuxiliarySuffix (n : Name) : Name := match n with
+    | .str parent s =>
+      -- Only strip if parent is not anonymous (to avoid stripping the entire name)
+      if !parent.isAnonymous && isAuxiliarySuffix s then parent else n
+    | _ => n
+  let cleanupName (n : Name) : Name :=
+    let n := stripAuxiliarySuffix n
+    if n == `_example then `example else n
+  decl.getTopLevelNames.map cleanupName
+
```

Comment:
While I agree that the feature is a useful QOL improvement, I am very reluctant to merge ad-hoc code like this in a very central location.

Maybe we need a general concept of “this is the user-facing declaration name that is related to the present one”, maybe useful in other places as well. But I don't want to think about designing that right away.

Maybe some other maintainer feels more generous here, but I’ll reject this in the current form, with “more thought needed”, I hope you understand.